compiler/luci/pass/src/VerifyQuantizedNodeType.h

```diff
@@ -0,0 +1,141 @@
+ * @tparam Btype Bias quantization type (e.g. For Q8, S32 is used)
+ */
+template <loco::DataType Qtype, loco::DataType Btype>
+class VerifyQuantizedNodeTypeBase : public luci::CircleNodeVisitor<bool>,
+                                    public VerifyQuantizedNodeType
+{
+public:
+  bool verify(luci::CircleNode *node) { return node->accept(this); }
+
+protected:
+  bool has_type(const loco::Node *node, loco::DataType dtype)
+  {
+    auto circle_node = loco::must_cast<const luci::CircleNode *>(node);
+    return circle_node->dtype() == dtype;
+  }
+
+private:
+  bool visit(const luci::CircleAdd *node);
+  bool visit(const luci::CircleArgMax *node);
+  bool visit(const luci::CircleAveragePool2D *node);
+  bool visit(const luci::CircleBatchToSpaceND *node);
+  bool visit(const luci::CircleCast *node);
+  bool visit(const luci::CircleConv2D *node);
+  bool visit(const luci::CircleConcatenation *node);
+  bool visit(const luci::CircleDepthToSpace *node);
+  bool visit(const luci::CircleDepthwiseConv2D *node);
+  bool visit(const luci::CircleDiv *node);
+  bool visit(const luci::CircleElu *node);
+  bool visit(const luci::CircleFloor *node);
+  bool visit(const luci::CircleFloorDiv *node);
+  bool visit(const luci::CircleFullyConnected *node);
+  bool visit(const luci::CircleGreater *node);
+  bool visit(const luci::CircleGreaterEqual *node);
+  bool visit(const luci::CircleInstanceNorm *node);
+  bool visit(const luci::CircleLocalResponseNormalization *node);
+  bool visit(const luci::CircleLogicalOr *node);
+  virtual bool visit(const luci::CircleLogistic *node) = 0;
+  bool visit(const luci::CircleMaxPool2D *node);
+  bool visit(const luci::CircleMean *node);
+  bool visit(const luci::CircleMirrorPad *node);
+  bool visit(const luci::CircleMul *node);
+  bool visit(const luci::CircleNotEqual *node);
+  bool visit(const luci::CircleOneHot *node);
+  bool visit(const luci::CirclePack *node);
+  bool visit(const luci::CirclePad *node);
+  bool visit(const luci::CirclePadV2 *node);
+  bool visit(const luci::CirclePRelu *node);
+  bool visit(const luci::CirclePow *node);
+  bool visit(const luci::CircleRelu *node);
+  bool visit(const luci::CircleReshape *node);
+  bool visit(const luci::CircleResizeBilinear *node);
+  bool visit(const luci::CircleResizeNearestNeighbor *node);
+  bool visit(const luci::CircleRsqrt *node);
+  bool visit(const luci::CircleSlice *node);
+  virtual bool visit(const luci::CircleSoftmax *node) = 0;
+  bool visit(const luci::CircleSpaceToBatchND *node);
+  bool visit(const luci::CircleSpaceToDepth *node);
+  bool visit(const luci::CircleSplit *node);
+  bool visit(const luci::CircleSplitOut *node);
+  bool visit(const luci::CircleSplitV *node);
+  bool visit(const luci::CircleSplitVOut *node);
+  bool visit(const luci::CircleSqrt *node);
+  bool visit(const luci::CircleStridedSlice *node);
+  virtual bool visit(const luci::CircleTanh *node) = 0;
+  bool visit(const luci::CircleTranspose *node);
```

Comment:
(optional) just optional; mixing pure virtual and implementation is OK but seems a little odd.
how about, (1) change `VerifyQuantizedNodeTypeBase` to `VerifyQuantizedNodeType` with all pure virtuals (2) introduce `VerifyQuantizedNodeTypeBase` with type neutral implementations only ?

Reply:
Thanks for your opinion and I agree that mixing them is not that "pretty" :)

But one of my original intention was to reduce duplicated codes.

#### Before
```cpp
bool visit(const luci::CircleTranspose *node)
{
  RETURN_FALSE_UNLESS(has_type(node, loco::DataType::U8))
  RETURN_FALSE_UNLESS(has_type(node->a(), loco::DataType::U8))
  RETURN_FALSE_UNLESS(has_type(node->perm(), loco::DataType::S32))
  return true;
}

bool visit(const luci::CircleTranspose *node)
{
  RETURN_FALSE_UNLESS(has_type(node, loco::DataType::S16))
  RETURN_FALSE_UNLESS(has_type(node->a(), loco::DataType::S16))
  RETURN_FALSE_UNLESS(has_type(node->perm(), loco::DataType::S32))
  return true;
}
```

#### After
```cpp
bool visit(const luci::CircleTranspose *node)
{
  RETURN_FALSE_UNLESS(has_type(node, Qtype))
  RETURN_FALSE_UNLESS(has_type(node->a(), Qtype))
  RETURN_FALSE_UNLESS(has_type(node->perm(), loco::DataType::S32))
  return true;
}
```

Unfortunately, three operation(Tanh, Softmax, Logistic) cannot be reduced as one function so I implemented as current code :(

---

The reason why I created `VerifyQuantizedNodeType` as sort of interface is for preparing when new kinds of quantization is introduced :)

```cpp
class VerifyNewQuantizationMethod : public VerifyQuantizedNodeType
{
//...
}
```

Reply:
Um... there seems some misunderstanding with my explanation.
I'll post a PR after this lands.

Reply:
Found out that removing pure virtuall will do